sim/scenarios/corrupt-rate/corrupt-rate-error-model.cc

```diff
@@ -0,0 +1,63 @@
+#include <cstdint>
+
+#include "corrupt-rate-error-model.h"
+#include "ns3/header.h"
+#include "ns3/packet.h"
+#include "ns3/ppp-header.h"
+#include "ns3/ipv4-header.h"
+#include "ns3/udp-header.h"
+
+using namespace std;
+
+NS_OBJECT_ENSURE_REGISTERED(CorruptRateErrorModel);
+
+TypeId CorruptRateErrorModel::GetTypeId(void) {
+    static TypeId tid = TypeId("CorruptRateErrorModel")
+        .SetParent<ErrorModel>()
+        .AddConstructor<CorruptRateErrorModel>()
+        ;
+    return tid;
+}
+ 
+CorruptRateErrorModel::CorruptRateErrorModel()
+    : rate(0), distr(0, 99) {
+    std::random_device rd;
+    rng = new std::mt19937(rd());
+}
+
+void CorruptRateErrorModel::DoReset(void) { }
+
+bool CorruptRateErrorModel::DoCorrupt(Ptr<Packet> p) {
+    if (distr(*rng) >= rate) return false;
+
+    const uint32_t p_len = p->GetSize();
+    uint8_t *buffer= new uint8_t[p_len];
+    p->CopyData(buffer, p->GetSize());
+    PppHeader ppp_hdr = PppHeader();
+    uint32_t ppp_hdr_len = p->RemoveHeader(ppp_hdr);
+    Ipv4Header ip_hdr = Ipv4Header();
+    uint32_t ip_hdr_len = p->RemoveHeader(ip_hdr);
+    UdpHeader udp_hdr = UdpHeader();
+    uint32_t udp_hdr_len = p->RemoveHeader(udp_hdr);
+    const uint32_t hdr_len = ppp_hdr_len + ip_hdr_len + udp_hdr_len;
+    std::uniform_int_distribution<> d(hdr_len, hdr_len + p->GetSize() - 1);
```

Comment:
There's not much value in corrupting the encrypted bytes, is it? That will trivially get caught by the decrypter. Perhaps limiting this to, say min (p->GetSize() - 1, 50) will limit this to the first 50 bytes, allowing more testing of QUIC header corruption. Granted much of that is encrypted, but hitting something in the QUIC header is still more interesting than hitting the rest of the packet. What do you think?

Reply:
That makes sense. Changed.